python/rascaline/rascaline/utils/clebsch_gordan/_density_correlations.py

```diff
@@ -0,0 +1,226 @@
+            correlation order ``n_correlations + 1``.
+        """
+        return self._density_correlations(
+            density,
+            angular_cutoff=angular_cutoff,
+            selected_keys=selected_keys,
+            skip_redundant=skip_redundant,
+            compute_metadata=False,
+        )
+
+    def _density_correlations(
+        self,
+        density: TensorMap,
+        angular_cutoff: Optional[int],
+        selected_keys: Optional[Labels],
+        skip_redundant: bool,
+        compute_metadata: bool,
+    ) -> TensorMap:
+        """
+        Computes the density auto-correlations.
+        """
+        # Prepare the tensors. Before the first iteration, both the 'left' and 'right'
+        # densities should have a "_1" suffix in the property names. The property names
+        # of the 'right' density is incremented by one at the start of each iteration.
+        density_correlations = index_property_names(density, 1)
+        density = index_property_names(density, 1)
+
+        # Perform iterative CG tensor products
+        new_lambda_names: List[str] = []
+        for i_correlation in range(self._n_correlations):
+
+            # Increment the density property dimension names
+            density = index_property_names(density, 1)
+
+            # Define new key dimension names for tracking intermediate correlations. The
+            # following conditional clause is a matter of convention. "l" indices track
+            # blocks of nu=1 densities, and "k" indices track blocks originating from
+            # higher order tensors.
+            if i_correlation == 0:
+                o3_lambda_1_name: str = f"l_{i_correlation + 1}"
+            else:
+                o3_lambda_1_name: str = f"k_{i_correlation + 1}"
+            o3_lambda_2_name: str = f"l_{i_correlation + 2}"
+            new_lambda_names.extend([o3_lambda_1_name, o3_lambda_2_name])
+
+            # Define the selected keys for the current iteration, applying the
+            # ``angular_cutoff`` if specified.
+            if (
+                i_correlation == self._n_correlations - 1 and selected_keys is None
+            ) or (i_correlation < self._n_correlations - 1):
+                if angular_cutoff is None:
+                    selected_keys_iteration = None
+                else:
+                    selected_keys_iteration: Labels = Labels(
+                        names=["o3_lambda"],
+                        values=_dispatch.int_range_like(
+                            0, angular_cutoff + 1, like=density.keys.values
+                        ).reshape(-1, 1),
+                    )
+            else:
+                assert i_correlation == self._n_correlations - 1
+                assert selected_keys is not None
+                selected_keys_iteration = selected_keys
+
+            # Compute CG tensor product
+            density_correlations = self._tensor_correlator._cg_tensor_product(
```

Comment:
this should call the public function:

```suggestion
            density_correlations = self._tensor_correlator.compute(
```